### PR TITLE
[fx] Escape all dot record metacharacters in FxGraphDrawer labels

### LINKED ISSUE #147884: `FxGraphDrawer` fails on `einsum` nodes
### 🐛 Describe the bug

I am trying to visualize a `torch.fx.GraphModule` using `torch.fx.passes.graph_drawer.FxGraphDrawer`.
If the graph module contains an `einsum` operation, I get a `bad label` error.

Here is an MWE to reproduce the problem:
```python
"""Cannot visualize `einsum` nodes with `torch.fx` graph drawer."""

from torch import einsum
from torch.fx import passes, symbolic_trace
from torch.nn import Module

# Setting this to `True` triggers the error.
# Everything works fine if set to `False`.
USE_EINSUM = True


class Square(Module):
    def forward(self, x):
        if USE_EINSUM:
            return einsum("a,a->a", x)
        else:
            return x**2


mod = symbolic_trace(Square())
g = passes.graph_drawer.FxGraphDrawer(mod, "einsum")
g.get_dot_graph().write_svg("einsum.svg")
``` 

Here is an example of the error:
```bash
b'Error: bad label format {name=%einsum|op_code=call_function\\n|target=torch.functional.einsum\\n|args=(a,a->a,)|num_users=1\\n}\n'
```

### Versions

```
PyTorch version: 2.4.0.post101
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.3.1 (x86_64)
GCC version: Could not collect
Clang version: 16.0.0 (clang-1600.0.26.6)
CMake version: version 3.31.0
Libc version: N/A

Python version: 3.9.16 (main, May 15 2023, 18:51:40)  [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2

Versions of relevant libraries:
[pip3] flake8==7.1.1
[pip3] flake8-bugbear==24.12.12
[pip3] flake8-comprehensions==3.16.0
[pip3] flake8-tidy-imports==4.11.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.0.1
[pip3] torch==2.4.0.post101
[conda] libtorch                  2.4.0           cpu_mkl_hdbae018_101    conda-forge
[conda] mkl                       2023.2.0         h54c2260_50500    conda-forge
[conda] numpy                     2.0.2                    pypi_0    pypi
[conda] numpy-base                2.0.1            py39h03d8c7d_1  
[conda] pytorch                   2.4.0           cpu_mkl_py39hd2dbf71_101    conda-forge
```

### PR DESCRIPTION
**Pull request by Chinmayrawat15**
Fixes #147884

### Root cause

`FxGraphDrawer` draws each node as a graphviz `record`, whose structure is defined by the characters `{`, `}`, `|`, `<` and `>`. Only `{` and `}` were escaped when graph content was interpolated into a label, so any other metacharacter reaching a label made dot reject the whole label with `Error: bad label format` and drop the node's contents from the rendered graph.

The reported case is an einsum equation: `torch.einsum("a,a->a", x)` puts a literal `>` into `args=(a,a->a,)`. The same failure is reachable from two more label fields that escaped nothing at all:

* a leaf module's `__constants__`, where a dict value renders as `{'a': 1}`
* the parsed stack trace, where both the source line and pseudo-filenames such as `<stdin>` are interpolated verbatim

### Fix

Escaping moves into one helper covering the full metacharacter set, applied to the content-bearing pieces only, so the `|` that separates record fields and the outer `{}` keep their structural meaning. The old code escaped the assembled string at the end instead; consequently the kwargs prefix/suffix braces move to their pre-escaped spelling. The emitted label for kwargs is unchanged, and since dot renders `\<` as a plain `<` in non-record shapes, `dot_graph_shape="box"` output is unchanged too.

### Test Plan

New `TestFXGraphDrawer` in `test/test_fx_passes.py`. The three `test_escapes_*` tests and `test_dot_renders_str_arg` all fail on main and pass with this change.

```
python test/test_fx_passes.py -k TestFXGraphDrawer
python test/test_fx_passes.py
python test/test_fx.py
python test/inductor/test_graph_transform_observer.py
python test/inductor/test_debug_graph_dump.py
```

Verified end to end against a real graphviz 15.1.1 by rendering each case to SVG. Five of the eight fail to render on main; all eight render with this change:

| case | before | after |
| --- | --- | --- |
| einsum equation arg (this issue) | fails | ok |
| `__constants__` dict / builtin repr | fails | ok |
| stack trace with `<stdin>` and `>` | fails | ok |
| kwargs record, multiple | ok | ok |
| kwargs record, exactly one | ok | ok |
| `dot_graph_shape="box"` | ok | ok |
| module with params + tensor_meta | fails | ok |
| same, `normalize_args=True` | fails | ok |

The last three rows are regression guards rather than new coverage.

### Note on CI coverage

`pydot` is not in `.ci/docker/requirements-ci.txt`, so the four pydot-gated tests will skip in CI and only the five `_escape_dot_label` subtests will actually run there. I followed the `HAS_PYDOT` / `HAS_DOT` gating already used by `test/inductor/test_graph_transform_observer.py`. I did not add `pydot` to the CI requirements myself because any change under `.ci/docker/` forces a full Docker image rebuild — happy to do it in a follow-up if you'd like this covered in CI.

This PR was prepared with the assistance of Claude Code.

### CODE REVIEW

**Comment by Chinmayrawat15** [comment]
Ready for review. Rebased onto current `main` just now — no conflicts, and nothing upstream has touched either file since I branched.

Short version: `FxGraphDrawer` renders each node as a graphviz `record`, whose structure is defined by `{ } | < >`. Only `{` and `}` were escaped, so an einsum equation (`torch.einsum("a,a->a", x)` → `args=(a,a->a,)`) makes dot reject the entire label. It still writes an SVG with that node's fields silently missing, which is why #147884 reads as a rendering quirk rather than an error. Two other label fields escaped nothing at all: a leaf module's `__constants__`, and the parsed stack trace — where `<stdin>`-style filenames come from any exec'd or notebook-traced module, and `parse_stack_trace=True` is what inductor's debug dumps use.

Evidence renders through a real graphviz 15.1.1 rather than asserting on strings: **3 of 8 cases render on main, 8 of 8 with this change.** Three of the eight pass before and after on purpose — they're regression guards for the kwargs path, which the fix re-plumbs.

One call I'd like a maintainer's opinion on, also flagged in the PR body: `pydot` isn't in `.ci/docker/requirements-ci.txt`, so 4 of the 9 new tests will skip in CI. I didn't add the dependency because every file under `.ci/docker/` is content-hashed and any edit forces a full image rebuild — happy to do it in a follow-up if you'd rather have the coverage.

The five workflows are sitting at `action_required` with zero runs, so CI hasn't started — could someone approve workflow runs?

I used Claude Code to help investigate and write this; I've reviewed and tested it myself, and the numbers above come from runs on my machine.

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

**Comment by zou3519** [comment]
@claude review this please and put the review inside a GitHub collapsible section

**Comment by Chinmayrawat15** [comment]
Thanks — worked through all nine. Eight are addressed in `cd2ae442`; two I'd push back on with measurements.

**Finding 1 is right and reachable.** A Windows path in a parsed stack trace renders as two mangled lines, because `\l` is read as the left-justify directive — and dot exits 0, so nothing surfaces:

```
label:   "{file=C:\lib\foo.py|num_users=1}"     exit 0, empty stderr
renders:  file=C:  /  ibfoo.py  /  num_users=1
```

Backslash is now escaped first. Your ordering warning was the load-bearing part: with content escaped, the one-item form's `.replace(r"\l", "")` would eat the `\l` out of `C:\\lib`. Rather than special-casing that, the one-item form now omits those breaks instead of stripping them back out afterwards. Both emitted labels are byte-identical to before, which the two new `test_kwargs_label_shape` cases pin exactly.

**Findings 2 and 5 — the mechanism is right, the exit code isn't.** On graphviz 15.1.1:

```
$ dot -Tsvg badlabel.dot > out.svg; echo $?
1
Error: bad label format {name=%einsum|args=(a,a->a,)}
$ wc -c out.svg
842
```

So the SVG *is* still written with the node name substituted for the fields, exactly as you describe — but dot exits **1**, not 0, and pydot's `create_svg()` raises `AssertionError` on a non-zero return code. I've corrected the code comment to describe the substitution (it did overstate "rejects the whole label"), and rewritten the note in `test_dot_renders_str_arg` to say why reaching the assertion matters without overstating it either. If you know the exit code differs on another graphviz version, I'll gate the claim rather than assert it.

**Finding 3** applied — `register_parameter` / `register_buffer` reject only `"."` and `""`, so `a|b` is a legal name that would split a field there. I left `buf_meta.name` as-is with a note, per your read that inductor only produces `buf0`-style names.

**Finding 4 is fair.** Those table rows were manual renders, and the PR body should have said so rather than listing them where they read as committed coverage. Added exact-label assertions for both kwargs branches — the one-item case takes the branch the diff re-plumbs — and for `dot_graph_shape="box"`.

**6, 7, 8, 9** applied: moved to `common_utils.TestCase`, reuse `graph_drawer.HAS_PYDOT` instead of re-deriving it with a wider `except`, replaced the no-op `nothing_to_escape` case with one mixing escaped and unescaped characters, hoisted the duplicated `Einsum`, and `_node_label` now reports what it actually found.

**On CI coverage — your point is sharper than mine was.** I checked it by stubbing `_escape_dot_label` to return its input unchanged: **12 of the 14 tests fail, including the pure-unit subtests that run in CI without pydot.** So a no-op helper would be caught; call sites being unwired still would not. Closing that needs `pydot` in `requirements-ci.txt`, which I'd rather land as a follow-up than bundle into this fix — happy to open it if you'd like.

14 tests, up from 9. Reverting only the backslash line fails exactly the two new Windows-path tests.

I used Claude Code to help investigate and write this; I verified each finding against the branch myself, and the numbers above are from runs on my machine.

**Review by zou3519 — COMMENTED** [review]
Sorry I don't think we want to accept changes here, as far as I know graph_drawer is not maintaned

**Comment by Chinmayrawat15** [comment]
Thanks for taking a look, and understood — happy to close this if that's still the call.

One thing I'd offer first, in case it changes the calculus. I went back to see why label changes here are expensive: #137726 refactored these labels to HTML-like and was reverted three days later for breaking internal components. This change is much narrower. It escapes `{}|<>` and `\`, so the only labels whose output changes are ones `dot` already rejects today with `Error: bad label format`. Nothing internal can be depending on output that currently errors, and labels without those characters come out byte-identical.

The one exception I should flag myself: a Windows path in a parsed stack trace currently renders mangled rather than erroring, because `\l` gets read as dot's left-justify directive. That case does change existing, wrong output.

It's a few lines in `_escape_dot_label` plus 14 tests, 12 of which fail if the escaping is stubbed out. The file still has 13 call sites — `partitioners.py`, `graph_transform_observer.py`, `splitter_base.py`, and Inductor's debug dumps — so it's not dead code, just unowned, which I take to be the real issue.

Either way, no hard feelings; I just didn't want to leave the smaller blast radius unstated.
